### PR TITLE
Java: Implement `APPEND` command

### PR DESCRIPTION
**Pull request by cyip10**
Implemented the append command. The SpotlessApply tool changed the TransactionTest.java imports to this import `static redis_request.RedisRequestOuterClass.RequestType.*;`

### CODE REVIEW

File: java/client/src/main/java/glide/api/commands/StringBaseCommands.java [hunk 1]
```diff
@@ -243,4 +243,30 @@ public interface StringBaseCommands {
...
      */
     CompletableFuture<String> getrange(String key, int start, int end);
+
+    /**
+     * Appends a value to a key. If `key` does not exist it is created and set as an empty string, so
+     * `APPEND` will be similar to `SET` in this special case.
+     *
+     * @see <a href="https://redis.io/docs/latest/commands/append/">redis.io</a> for details.
+     * @param key The <code>key</code> to which the value will be appended.
+     * @param value The value to append
```
Comment (Yury-Fridlyand) [inline java/client/src/main/java/glide/api/commands/StringBaseCommands.java:253]:
```suggestion
     * @param value The value to append.
```

File: java/client/src/main/java/glide/api/commands/StringBaseCommands.java [hunk 2]
```diff
@@ -243,4 +243,30 @@ public interface StringBaseCommands {
...
+     * @return Response from Redis. The length of the string after appending the value.
+     * @example
+     *     <pre>{@code
+     * String value = client.append("key", "value").get();
+     * assert value.equals(5);
+     * Indicates that "Hello" has been appended to the value of <code>key</code>, which was initially empty,
+     * resulting in a new value of "Hello" with a length of 5 - similar to the set operation.
```
Comment (Yury-Fridlyand) [inline java/client/src/main/java/glide/api/commands/StringBaseCommands.java:260]: The doc should contain a compilable code sample
```suggestion
     * // Indicates that "Hello" has been appended to the value of <code>key</code>, which was initially empty,
     * // resulting in a new value of "Hello" with a length of 5 - similar to the set operation.
```

File: java/client/src/main/java/glide/api/commands/StringBaseCommands.java [hunk 3]
```diff
@@ -243,4 +243,30 @@ public interface StringBaseCommands {
...
+     *
+     * String value = client.append("key", " world").get();
```
Comment (Yury-Fridlyand) [inline java/client/src/main/java/glide/api/commands/StringBaseCommands.java:262]: In java we ~usually~ always add only one example

File: java/client/src/main/java/glide/api/models/BaseTransaction.java [hunk 4]
```diff
@@ -2862,4 +2862,18 @@ protected ArgsArray buildArgs(String... stringArgs) {
 
         return commandArgs.build();
     }
+
+    /**
+     * Appends a value to a key.
```
Comment (Yury-Fridlyand) [inline java/client/src/main/java/glide/api/models/BaseTransaction.java:2867]: Copy full doc from the interface here (except example)

File: java/client/src/main/java/glide/api/models/BaseTransaction.java [hunk 5]
```diff
@@ -2862,4 +2862,18 @@ protected ArgsArray buildArgs(String... stringArgs) {
...
+     *
+     * @see <a href="https://redis.io/commands/append">redis.io</a> for more details.
+     * @param key The key to which the value will be appended.
+     * @param value The value to append.
+     * @return The length of the string after appending the <code>value</code>.
```
Comment (Yury-Fridlyand) [inline java/client/src/main/java/glide/api/models/BaseTransaction.java:2872]:
```suggestion
     * @return Command Respose - The length of the string after appending the <code>value</code>.
```

File: java/client/src/main/java/glide/api/commands/StringBaseCommands.java [hunk 6]
```diff
@@ -243,4 +243,30 @@ public interface StringBaseCommands {
      * }</pre>
      */
     CompletableFuture<String> getrange(String key, int start, int end);
+
+    /**
+     * Appends a value to a key. If `key` does not exist it is created and set as an empty string, so
+     * `APPEND` will be similar to `SET` in this special case.
```
Comment (Yury-Fridlyand) [inline java/client/src/main/java/glide/api/commands/StringBaseCommands.java:249]: Not sure that we need such detailed doc, but if you refer to `set`, add a link to that command. Javadoc renderer will make it clickable.
```suggestion
     * Appends a value to a key. If `key` does not exist it is created and set as an empty string, so
     * `APPEND` will be similar to {@see #set} in this special case.
```
(I hope I didn't mistake in the syntax)

File: java/client/src/test/java/glide/api/models/TransactionTests.java [hunk 7]
```diff
@@ -20,116 +20,7 @@
...
-import static redis_request.RedisRequestOuterClass.RequestType.ZScore;
-import static redis_request.RedisRequestOuterClass.RequestType.ZUnion;
-import static redis_request.RedisRequestOuterClass.RequestType.ZUnionStore;
-import static redis_request.RedisRequestOuterClass.RequestType.Zadd;
-import static redis_request.RedisRequestOuterClass.RequestType.Zcard;
-import static redis_request.RedisRequestOuterClass.RequestType.Zcount;
-import static redis_request.RedisRequestOuterClass.RequestType.Zrange;
-import static redis_request.RedisRequestOuterClass.RequestType.Zrank;
-import static redis_request.RedisRequestOuterClass.RequestType.Zrem;
+import static redis_request.RedisRequestOuterClass.RequestType.*;
```
Comment (Yury-Fridlyand) [inline java/client/src/test/java/glide/api/models/TransactionTests.java:23]: We don't do star imports [yet]. Probably, IDE made this for you - try to reconfigure and revert.

File: java/integTest/src/test/java/glide/SharedCommandTests.java [hunk 8]
```diff
@@ -3390,4 +3390,16 @@ public void geopos(BaseClient client) {
...
+    @SneakyThrows
+    @ParameterizedTest(autoCloseArguments = false)
+    @MethodSource("getClients")
+    public void append(BaseClient client) {
+        String key1 = UUID.randomUUID().toString();
+        String value = String.valueOf(UUID.randomUUID());
+
+        assertEquals(value.length(), client.append(key1, value).get());
+        assertEquals(value.length() * 2L, client.append(key1, value).get());
+        assertEquals(value + value, client.get(key1).get());
```
Comment (Yury-Fridlyand) [inline java/integTest/src/test/java/glide/SharedCommandTests.java:3403]: Add a test where you apply `append` to a key which is empty and a test where key does not store a string (but stores a hash, for example).

**Review by Yury-Fridlyand — COMMENTED** [review]
CI is red - why? spotless?

File: java/client/src/main/java/glide/api/commands/StringBaseCommands.java (hunk 2 above)
Comment (cyip10) [reply java/client/src/main/java/glide/api/commands/StringBaseCommands.java:260]: ohh so adding // will just comment out the comment code

File: java/client/src/main/java/glide/api/commands/StringBaseCommands.java (hunk 3 above)
Comment (cyip10) [reply java/client/src/main/java/glide/api/commands/StringBaseCommands.java:262]: @Yury-Fridlyand i copied over the python code. is it better if I just remove one example?

File: java/client/src/main/java/glide/api/models/BaseTransaction.java (hunk 5 above)
Comment (cyip10) [reply java/client/src/main/java/glide/api/models/BaseTransaction.java:2872]: got it

File: java/client/src/main/java/glide/api/commands/StringBaseCommands.java (hunk 1 above)
Comment (cyip10) [reply java/client/src/main/java/glide/api/commands/StringBaseCommands.java:253]: got it

File: java/client/src/main/java/glide/api/commands/StringBaseCommands.java (hunk 6 above)
Comment (cyip10) [reply java/client/src/main/java/glide/api/commands/StringBaseCommands.java:249]: got it

**Comment by acarbonetto** [comment]
CI is failing for a compile error.  
``` 
/home/runner/work/glide-for-redis/glide-for-redis/java/client/src/main/java/glide/api/models/BaseTransaction.java:2876: error: cannot find symbol

> Task :client:compileJava
        protobufTransaction.addCommands(buildCommand(Append, commandArgs));
                                                     ^
  symbol:   variable Append
  location: class BaseTransaction<T>
  where T is a type-variable:
    T extends BaseTransaction<T> declared in class BaseTransaction
Note: /home/runner/work/glide-for-redis/glide-for-redis/java/client/src/main/java/glide/api/models/Script.java uses or overrides a deprecated API.
Note: Recompile with -Xlint:deprecation for details.
1 error
```

File: java/client/src/main/java/glide/api/commands/StringBaseCommands.java [hunk 9]
```diff
@@ -243,4 +243,30 @@ public interface StringBaseCommands {
...
+     *
+     * @see <a href="https://redis.io/docs/latest/commands/append/">redis.io</a> for details.
+     * @param key The <code>key</code> to which the value will be appended.
```
Comment (acarbonetto) [inline java/client/src/main/java/glide/api/commands/StringBaseCommands.java:252]:
```suggestion
     * @param key The key of the string. 
```

File: java/client/src/main/java/glide/api/commands/StringBaseCommands.java (hunk 2 above)
Comment (acarbonetto) [reply java/client/src/main/java/glide/api/commands/StringBaseCommands.java:260]: Yes.  This should be code that can be copy-pasted for a user to use.

File: java/client/src/main/java/glide/api/commands/StringBaseCommands.java [hunk 10]
```diff
@@ -243,4 +243,30 @@ public interface StringBaseCommands {
...
+     * @example
+     *     <pre>{@code
+     * String value = client.append("key", "value").get();
+     * assert value.equals(5);
+     * Indicates that "Hello" has been appended to the value of <code>key</code>, which was initially empty,
+     * resulting in a new value of "Hello" with a length of 5 - similar to the set operation.
+     *
+     * String value = client.append("key", " world").get();
+     * assert value.equals(11);
+     * Indicates that " world" has been appended to the value of <code>key</code>, resulting in a new value of "Hello world" with a length of 11.
```
Comment (acarbonetto) [inline java/client/src/main/java/glide/api/commands/StringBaseCommands.java:264]:
```suggestion
     * // Indicates that " world" has been appended to the value of <code>key</code>, resulting in a new value of "Hello world" with a length of 11.
```

File: java/client/src/main/java/glide/api/commands/StringBaseCommands.java (hunk 6 above)
Comment (acarbonetto) [reply java/client/src/main/java/glide/api/commands/StringBaseCommands.java:249]:
```
     * Appends <code>value</code> to the string at <code>key</code>. If <code>key</code> does not exist it is created and set as an empty string, so
     * `APPEND` will be similar to `SET` in this special case.
```
Use code tags when referring to arguments and Java types.

File: java/integTest/src/test/java/glide/SharedCommandTests.java (hunk 8 above)
Comment (acarbonetto) [reply java/integTest/src/test/java/glide/SharedCommandTests.java:3403]: > Add a test where you apply append to a key which is empty

That's line 3401.

File: java/integTest/src/test/java/glide/TransactionTestUtilities.java [hunk 11]
```diff
@@ -199,6 +199,9 @@ public static BaseTransaction<?> transactionTest(BaseTransaction<?> baseTransact
                 .pfmerge(hllKey3, new String[] {hllKey1, hllKey2})
                 .pfcount(new String[] {hllKey3});
 
+        baseTransaction.append(key1, value1);
```
Comment (acarbonetto) [inline java/integTest/src/test/java/glide/TransactionTestUtilities.java:202]: can you move this test up to just after the strlen command (line 58)?  That way its the same test in PYthon and Java.

File: java/client/src/main/java/glide/api/commands/StringBaseCommands.java (hunk 3 above)
Comment (Yury-Fridlyand) [reply java/client/src/main/java/glide/api/commands/StringBaseCommands.java]: Yes. All python client examples (where they are) are more verbose.

File: java/client/src/main/java/glide/api/commands/StringBaseCommands.java [hunk 12]
```diff
@@ -243,4 +243,21 @@ public interface StringBaseCommands {
...
+     * case.
+     *
+     * @see <a href="https://redis.io/docs/latest/commands/append/">redis.io</a> for details.
+     * @param key The key of the string.
+     * @param value The value to append.
+     * @return Response from Redis. The length of the string after appending the value.
+     * @example
+     *     <pre>{@code
+     * String value = client.append("key", "value").get();
+     * assert value.equals(5);
```
Comment (tjzhang-BQ) [inline java/client/src/main/java/glide/api/commands/StringBaseCommands.java:259]: nit: should it be 5L?

File: java/client/src/main/java/glide/api/models/BaseTransaction.java [hunk 13]
```diff
@@ -2862,4 +2863,19 @@ protected ArgsArray buildArgs(String... stringArgs) {
 
         return commandArgs.build();
     }
+
+    /**
+     * Appends a value to a key. If `key` does not exist it is created and set as an empty string, so
+     * `APPEND` will be similar to {@see #set} in this special case.
+     *
+     * @see <a href="https://redis.io/docs/latest/commands/append/">redis.io</a> for details.
+     * @param key The <code>key</code> to which the value will be appended.
```
Comment (Yury-Fridlyand) [inline java/client/src/main/java/glide/api/models/BaseTransaction.java:2872]:
```suggestion
     * @param key The key of the string.
```
docs should match with the interface

File: java/client/src/main/java/glide/api/models/BaseTransaction.java [hunk 14]
```diff
@@ -2862,4 +2863,19 @@ protected ArgsArray buildArgs(String... stringArgs) {
 
         return commandArgs.build();
     }
+
+    /**
+     * Appends a value to a key. If `key` does not exist it is created and set as an empty string, so
```
Comment (Yury-Fridlyand) [inline java/client/src/main/java/glide/api/models/BaseTransaction.java:2868]: Copy updated doc from interface here please

File: java/client/src/main/java/glide/api/commands/StringBaseCommands.java [hunk 15]
```diff
@@ -243,4 +243,21 @@ public interface StringBaseCommands {
...
+     * created and set as an empty string, so `APPEND` will be similar to {@see #set} in this special
+     * case.
+     *
+     * @see <a href="https://redis.io/docs/latest/commands/append/">redis.io</a> for details.
+     * @param key The key of the string.
+     * @param value The value to append.
+     * @return Response from Redis. The length of the string after appending the value.
+     * @example
+     *     <pre>{@code
+     * String value = client.append("key", "value").get();
```
Comment (Yury-Fridlyand) [inline java/client/src/main/java/glide/api/commands/StringBaseCommands.java:258]:
```suggestion
     * Long value = client.append("key", "value").get();
```

File: java/client/src/test/java/glide/api/models/TransactionTests.java (hunk 7 above)
Comment (acarbonetto) [reply java/client/src/test/java/glide/api/models/TransactionTests.java:23]: Can we revert this or is spotless enforcing?

File: java/integTest/src/test/java/glide/TransactionTestUtilities.java [hunk 16]
```diff
@@ -46,6 +46,7 @@ public class TransactionTestUtilities {
     private static final String field1 = UUID.randomUUID().toString();
     private static final String field2 = UUID.randomUUID().toString();
     private static final String field3 = UUID.randomUUID().toString();
+    private static final String key10 = "{key}" + UUID.randomUUID();
```
Comment (acarbonetto) [inline java/integTest/src/test/java/glide/TransactionTestUtilities.java:49]: we don't need to create a new key

File: java/integTest/src/test/java/glide/TransactionTestUtilities.java [hunk 17]
```diff
@@ -56,6 +57,8 @@ public static BaseTransaction<?> transactionTest(BaseTransaction<?> baseTransact
 
         baseTransaction.set(key2, value2, SetOptions.builder().returnOldValue(true).build());
         baseTransaction.strlen(key2);
+        baseTransaction.append(key10, value1);
+        baseTransaction.get(key10);
```
Comment (acarbonetto) [inline java/integTest/src/test/java/glide/TransactionTestUtilities.java:61]: According to the Python test, they use key2 here. And double the length of the string there. 

Note: then they have a mset call - which you could include. 

```
        baseTransaction.append(key2, value1);
        baseTransaction.mset(Map.of(key1, value2, key2, value1));
```

and the result... 
```
            Long.valueOf(value1.length() * 2), // append(key2, value1)
            OK, // mset(Map.of(key1, value2, key2, value1)
```

https://github.com/Bit-Quill/glide-for-redis/blob/main/python/python/tests/test_transaction.py#L72-L73

File: java/integTest/src/test/java/glide/TransactionTestUtilities.java (hunk 16 above)
Comment (cyip10) [reply java/integTest/src/test/java/glide/TransactionTestUtilities.java:49]: @acarbonetto when i tried using an existing key, it didn't give me the output I wanted so I made a new key.

Comment (acarbonetto) [reply java/integTest/src/test/java/glide/TransactionTestUtilities.java:49]: for transactions, you only need to include a very basic test case.  

can you try my suggestion below and see if the tests pass?  You could even run append with an empty string (which shouldn't change the result) and that would be good enough ;)

Comment (cyip10) [reply java/integTest/src/test/java/glide/TransactionTestUtilities.java:49]: yup will do

File: java/integTest/src/test/java/glide/TransactionTestUtilities.java (hunk 17 above)
Comment (cyip10) [reply java/integTest/src/test/java/glide/TransactionTestUtilities.java:61]: @acarbonetto when I add in your new implementation, it causes a test in the TransactionTests.java to break because it's now expecting a different value than what it initially had since you wrote this test before I wrote the append method. I think it's best to add another key to avoid any of these test conflicts. Thoughts?

File: java/client/src/main/java/glide/api/models/BaseTransaction.java [hunk 18]
```diff
@@ -2862,4 +2863,20 @@ protected ArgsArray buildArgs(String... stringArgs) {
...
+
+    /**
+     * Appends a <code>value</code> to a <code>key</code>. If <code>key</code> does not exist it is
+     * created and set as an empty string, so <code>APPEND</code> will be similar to {@see #set} in
+     * this special case.
+     *
+     * @see <a href="https://redis.io/docs/latest/commands/append/">redis.io</a> for details.
+     * @param key The key of the string.
+     * @param value The value to append.
+     * @return The length of the string after appending the value.
```
Comment (tjzhang-BQ) [inline java/client/src/main/java/glide/api/models/BaseTransaction.java:2875]:
```suggestion
     * @return Command Response - The length of the string after appending the value.
```

File: java/integTest/src/test/java/glide/TransactionTestUtilities.java [hunk 19]
```diff
@@ -56,7 +56,9 @@ public static BaseTransaction<?> transactionTest(BaseTransaction<?> baseTransact
 
         baseTransaction.set(key2, value2, SetOptions.builder().returnOldValue(true).build());
         baseTransaction.strlen(key2);
-        baseTransaction.customCommand(new String[] {"MGET", key1, key2});
+        baseTransaction.append(key2, value2);
+        baseTransaction.mset(Map.of(key2, value2));
```
Comment (Yury-Fridlyand) [inline java/integTest/src/test/java/glide/TransactionTestUtilities.java:60]: Why you do `mset`? to restore the value?
You can set `set` too, but this is not a mistake.

**Review by Yury-Fridlyand — APPROVED** [review]
Approved, assuming other comments are addressed

File: java/integTest/src/test/java/glide/SharedCommandTests.java [hunk 20]
```diff
@@ -3390,4 +3390,27 @@ public void geopos(BaseClient client) {
...
+        String value1 = String.valueOf(UUID.randomUUID());
+        String key2 = UUID.randomUUID().toString();
+        String key3 = UUID.randomUUID().toString();
+
+        assertEquals(value1.length(), client.append(key1, value1).get());
+        assertEquals(value1.length() * 2L, client.append(key1, value1).get());
+        assertEquals(value1 + value1, client.get(key1).get());
+
+        // Non Existing Key
+        assertEquals(1L, client.append(key2, "b").get());
```
Comment (acarbonetto) [inline java/integTest/src/test/java/glide/SharedCommandTests.java:3408]: we can remove this test.  Line 3403 does the same thing.

File: java/integTest/src/test/java/glide/SharedCommandTests.java [hunk 21]
```diff
@@ -3390,4 +3390,27 @@ public void geopos(BaseClient client) {
...
+    @SneakyThrows
+    @ParameterizedTest(autoCloseArguments = false)
+    @MethodSource("getClients")
+    public void append(BaseClient client) {
+        String key1 = UUID.randomUUID().toString();
+        String value1 = String.valueOf(UUID.randomUUID());
+        String key2 = UUID.randomUUID().toString();
+        String key3 = UUID.randomUUID().toString();
+
+        assertEquals(value1.length(), client.append(key1, value1).get());
```
Comment (acarbonetto) [inline java/integTest/src/test/java/glide/SharedCommandTests.java:3403]: Maybe move the comment from 3407 to here?  
```
// Append on non-existing string is like SET
```

File: java/integTest/src/test/java/glide/TransactionTestUtilities.java (hunk 19 above)
Comment (acarbonetto) [reply java/integTest/src/test/java/glide/TransactionTestUtilities.java:60]: it's more similar to the Python tests like this.

**Review by acarbonetto — COMMENTED** [review]
Almost there!

File: java/integTest/src/test/java/glide/SharedCommandTests.java [hunk 22]
```diff
@@ -3390,4 +3390,27 @@ public void geopos(BaseClient client) {
...
+        assertEquals(value1.length() * 2L, client.append(key1, value1).get());
+        assertEquals(value1 + value1, client.get(key1).get());
```
Comment (GumpacG) [inline java/integTest/src/test/java/glide/SharedCommandTests.java:3405]:
```suggestion
        assertEquals(value1.concat(value1), client.get(key1).get());
```

File: java/integTest/src/test/java/glide/SharedCommandTests.java (hunk 20 above)
Comment (GumpacG) [reply java/integTest/src/test/java/glide/SharedCommandTests.java:3408]: It's for this line https://github.com/Bit-Quill/glide-for-redis/pull/281/files#r1605442233. `key3` can be removed as `key2` can be used for the test for wrong type.

File: java/integTest/src/test/java/glide/TransactionTestUtilities.java (hunk 19 above)
Comment (GumpacG) [reply java/integTest/src/test/java/glide/TransactionTestUtilities.java:60]: Would it not be better to just at the end of the test but before `flush`? That way we don't add to unnecessary checks and it also won't modify keys needed for other tests.

**Review by acarbonetto — APPROVED** [review]
